x/qgb/abci.go

```diff
@@ -113,3 +122,108 @@ func handleValsetRequest(ctx sdk.Context, k keeper.Keeper) {
 		}
 	}
 }
+
+func PruneIfNeeded(ctx sdk.Context, k keeper.Keeper) {
+	if !k.CheckLatestAttestationNonce(ctx) ||
+		k.CheckLatestAttestationNonce(ctx) && k.GetLatestAttestationNonce(ctx) <= AttestationPruningThreshold {
+		// If the attestations nonce hasn't been initialized yet, or the nonce is not greater than the minimum
+		// number of attestations, we don't need to prune anything
+		return
+	}
+
+	lastAvailableNonce := k.GetLastPrunedAttestationNonce(ctx) + 1
+	lastAttestationInStore, found, err := k.GetAttestationByNonce(ctx, lastAvailableNonce)
+	if err != nil {
+		panic(err)
+	}
+	if !found {
+		panic(fmt.Sprintf("couldn't find attestation %s for pruning", lastAttestationInStore))
+	}
+	var lastNonceHeight uint64
+	switch lastAttestationInStore.Type() {
+	case types.DataCommitmentRequestType:
+		dc, ok := lastAttestationInStore.(*types.DataCommitment)
+		if !ok {
+			panic("couldn't cast data commitment type")
+		}
+		lastNonceHeight = dc.EndBlock
+	case types.ValsetRequestType:
+		vs, ok := lastAttestationInStore.(*types.Valset)
+		if !ok {
+			panic("couldn't cast valset type")
+		}
+		lastNonceHeight = vs.Height
+	}
+	lastUnbondingHeight := k.GetLastUnBondingBlockHeight(ctx)
+	if lastNonceHeight == lastUnbondingHeight {
+		// we don't need to do anything, we want to keep attestations up to the last unbonding height
+		return
+	}
+	if lastNonceHeight > lastUnbondingHeight {
+		// checking if it's the initial case following the startup of the chain
+		if lastNonceHeight == 1 && lastUnbondingHeight == 0 {
+			return
+		}
+		// we should never hit this case, since we will keep the attestations up to the last unbonding height
+		panic("missing attestations up to the unbonding height")
+	}
+	// now we have attestations before the unbonding height, we should check whether we need to prune them
+	// or not yet
+	latestAttestationNonce := k.GetLatestAttestationNonce(ctx)
+	if latestAttestationNonce-lastAvailableNonce+1 <= AttestationPruningThreshold {
+		// we don't need to prune as we still have room to store more attestations
+		return
+	}
```

Comment:
The one we have at the top is different, since it checks the special case of the time before pruning anything.
The first one checks if the nonces are still under the threshold,  then there is no need to waste gas anymore because we  know we generally want to keep that.

This check here is after the nonces are beyond the threshold and we need to start pruning.

The first check is more like a gas optimization

Reply:
ahh I think I see thanks for explaining